### PR TITLE
[DeadCode] Skip clone and new self on RemoveUnusedPrivatePropertyRector

### PR DESCRIPTION
**Pull request by samsonasik**
Given the following code:

```php
class SkipCloneSelf
{
    private $var;

    public function __construct($var)
    {
        $this->var = $var;
    }

    public function run()
    {
        $obj = clone $this;
        echo $obj->var;
    }

    public function run2($var)
    {
        $obj = new self($var);
        echo $obj->var;
    }
}
```

It produce:

```diff
-    private $var;
-
-    public function __construct($var)
+    public function __construct()
     {
-        $this->var = $var;
     }
```

which should be skipped, ref https://3v4l.org/jjNe3

### CODE REVIEW

**Comment by samsonasik** [comment]
Fixed 🎉

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: rules/Removing/NodeManipulator/ComplexNodeRemover.php [hunk 1]
```diff
@@ -57,46 +62,122 @@ public function removePropertyAndUsages(Property $property, array $classMethodNa
...
+    private function processRemovePropertyAssigns(array $assigns): void
+    {
+        foreach ($assigns as $assign) {
+            // remove assigns
+            $this->assignRemover->removeAssignNode($assign);
+            $this->removeConstructorDependency($assign);
+        }
+    }
+
+    private function isPropertyNameInNewCurrentClassNameSelfClone(string $propertyName, ?ClassLike $classLike): bool
```
Comment (TomasVotruba) [inline rules/Removing/NodeManipulator/ComplexNodeRemover.php]: Could you extract this to own service? The logic seems quite complex

Comment (samsonasik) [reply rules/Removing/NodeManipulator/ComplexNodeRemover.php]: implemented, I moved to new service `ForbiddenPropertyRemovalAnalyzer` https://github.com/rectorphp/rector-src/pull/1215/commits/12bf6944360c1ca17b855f6a56290f11a8429b87

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by TomasVotruba** [comment]
Thank you :+1:

**Pull request merged**